### PR TITLE
BUG: allow None as name in multi-index during join

### LINKED ISSUE #58721: BUG: joining dataframes with multi-index and None index label results in AssertionError
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
df1 = pd.DataFrame({"A": [1]}, index=pd.MultiIndex.from_tuples([(3, 3)], names=["X", None]))
df2 = pd.DataFrame({"B": [1]}, index=pd.MultiIndex.from_tuples([(3, 3)], names=[None, "X"]))
df1.join(df2)
```

### Issue Description

Fails with error: AssertionError: Length of order must be same as number of levels (2), got 1

### Expected Behavior

Should not result in AssertionError

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.9.18.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 23.4.0
Version               : Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000
machine               : x86_64
processor             : i386
byteorder             : little
LC_ALL                : None
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.1
numpy                 : 1.26.4
pytz                  : 2023.3.post1
dateutil              : 2.8.2
setuptools            : 68.2.2
pip                   : 23.3.1
Cython                : None
pytest                : 7.4.4
hypothesis            : None
sphinx                : 5.0.2
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.3
IPython               : 8.15.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.2
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : 15.0.1
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.13.0
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : 2.4.1
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by matiaslindgren**
- [x] closes #58721
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Another possible fix is to throw if `MultiIndex.names` contains `None`.
The exception message could say something like "empty index names must be empty strings, not None".

### CODE REVIEW

File: doc/source/whatsnew/v3.0.0.rst [hunk 1]
```diff
@@ -28,6 +28,7 @@ enhancement2
 
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
+- :attr:`MultiIndex.names` containing ``None`` no longer throws ``AssertionError`` during join (:issue:`58721`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v3.0.0.rst:31]: Can you move the note to the `Reshaping` section of bugfixes.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
This is looking good to me, but we were explicit about not allowing `None` in the past and I'm not sure why. Traced it back to:

- issue: https://github.com/pandas-dev/pandas/issues/3662
- commit: https://github.com/pandas-dev/pandas/commit/dd79e5510368723e17955a339a089727af3f4ea6

Don't see any discussion about `None`. Can you take a look @mroeschke

**Comment by matiaslindgren** [comment]
Automatic codeowner review requests were unintended, sorry. I accidentally pushed commits from my fork to this branch and that touched files unrelated to this PR. Force-pushed them away.

File: doc/source/whatsnew/v3.0.0.rst [hunk 2]
```diff
@@ -654,6 +654,7 @@ Groupby/resample/rolling
 
 Reshaping
 ^^^^^^^^^
+- :attr:`MultiIndex.names` containing ``None`` no longer throws ``AssertionError`` during join (:issue:`58721`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:657]:
```suggestion
- Bug in :meth:`DataFrame.join` when a :class:`DataFrame` with a :class:`MultiIndex` would raise an ``AssertionError`` when :attr:`MultiIndex.names` contained ``None``. (:issue:`58721`)
```

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by mroeschke** [comment]
pre-commit.ci autofix

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @matiaslindgren

**Pull request merged**